### PR TITLE
Replace the side inserter by an inserter with shortcuts on empty paragraphs

### LINKED ISSUE #4951: Revisit the trailing inserter
As of the time of this writing, `master` features a new interface for inserting content in the context of the canvas. It has a side-inserter, no sibling inserter and a trailing inserter consisting only of a plain textfield. This is part of a cage match to simplify in-canvas insertion (#4669). 

Having used this interface since it was merged in after the 2.1 release, we've had time to test it, even if it hasn't yet been released in the plugin. Already, feedback (including on the merge PR, #4539) has made it clear that aspects of the old interface are missing in the new, and aspects of the new weren't working as intended. This effectively closes #4669. As such, let's revert the side inserter, restore the sibling inserter, and revisit the trailing inserter.

Here is a mockup for an evolution of the trailing inserter:

trailing-inserter-v2

This is essentially a text input field giving you the usual access to either keep writing, or to use the Slash command. It restores recently used blocks on the right, but with Image and the inserter permanently added on the right. When you move the mouse towards these, or tab into them, the contrast increases. 

This is inspired by the Dropbox Paper interface:

paper

As a reminder, here's what's in the 2.1 version of the plugin:

☝️ this is mostly what the trailing inserter has looked like since June 2017, [where it looked like this](https://cloudup.com/cHelDRmdj11).

CC: @karmatosed @mtias @aduth @youknowriad

### PR DESCRIPTION
**Pull request by youknowriad**
closes #4951 

This PR updates the side inserter. Instead of having a side inserter at the right of each block we show a quick inserter with shortcuts for the most frequently used blocks on empty paragraphs.

Combined with the default block appender, this allows a nice writing flow, where you click the empty area at the bottom of a post to create a new empty paragraph which will trigger this side inserter.

### CODE REVIEW

**Comment by jasmussen** [comment]
Amazingly fast work. Also closes #4954. Taking a look.

**Comment by jasmussen** [comment]
This is already feeling much better than anything we've done in the past.

**Comment by paulwilde** [comment]
I think the hover styles should be rounded like they are in menu at the top. Seems to be 4px.

screen shot 2018-02-08 at 19 23 47

File: editor/store/selectors.js
Comment (noisysocks) [inline editor/store/selectors.js]: I'm wondering if we should we get rid of `MAX_RECENT_BLOCKS` and force callers to just always pass in a maximum.

File: editor/store/selectors.js
Comment (noisysocks) [inline editor/store/selectors.js]: Would `getFrequentInserterItems` be a better name for this? Then there would be a nice duality between this function and `getRecentInserterItems`.

File: editor/store/test/selectors.js
Comment (noisysocks) [inline editor/store/test/selectors.js]: Typo: s/8/4/

File: editor/components/inserter-with-shortcuts/index.js
Comment (noisysocks) [inline editor/components/inserter-with-shortcuts/index.js]: Should this be named `itemsWithoutDefaultBlock`?

File: editor/components/inserter-with-shortcuts/index.js
Comment (noisysocks) [inline editor/components/inserter-with-shortcuts/index.js]: We should pass in the proper `enabledBlockTypes` setting instead of `true` here.

File: editor/components/inserter-with-shortcuts/index.js
Comment (noisysocks) [inline editor/components/inserter-with-shortcuts/index.js]: `! item.initialAttributes` will _always_ evaluate to false. I think this should be:

```js
const itemsWithoutDefaultBlock = filter( items, ( item ) =>
    item.name !== getDefaultBlockName() || Object.keys( item.initialAttributes ).length > 0
).slice( 0, 2 );
```

Using `without` and `isEmpty` would help make it read more naturally:

```js
const itemsWithoutDefaultBlock = without( items, ( item ) =>
    item.name === getDefaultBlockName() && isEmpty( item.initialAttributes )
).slice( 0, 2 );
```

**Review by noisysocks — DISMISSED** [review]

**Comment by noisysocks** [comment]
Just realised this has the `In Progress` label. Apologies for butting in so early with code comments 🙂 —it's looking really good!

**Comment by youknowriad** [comment]
@noisysocks No problem, it's always good to have feedback. And I would have missed most of these ;). Thanks

**Comment by mtias** [comment]
@youknowriad one more thing to look at is the blue inserter line when you click on the + button — it should not be there as you'd be replacing the existing block instead:

image

Also, related to 4492, none of the block UI should be showing when clicking on the + button.

**Comment by mellis32** [comment]
Will the inserter shortcuts be grayed out until hovered as in #4951 and Dropbox Paper? I like the subtly of that apperance, which puts the focus back on the content.

**Comment by youknowriad** [comment]
I made some tweaks here, this should be closer to merge now. I'd love to have some input.

**Comment by mtias** [comment]
@youknowriad this is working really well for me. If there's one last thing to add, it'd be to try not showing the block outline, the paragraph toolbar, and the ellipsis menu when the paragraph block is empty.

**Comment by youknowriad** [comment]
Got a DM approval for this one. Merging to be ready for the release.

There's still an edge-case where the `onChange` is not triggered properly (clicking `backspace` while there's a selection, I guess TinyMCE should trigger `input` in this case. I'm considering this minor and should not be blocking this PR)

**Comment by jasmussen** [comment]
🎉🔥

**Comment by aduth** [comment]
I really like this iteration.

One issue I've noticed: It's not possible to expand the "More Options" block ellipsis menu with an empty default block.

Potentially related to https://github.com/WordPress/gutenberg/pull/4872#discussion_r167594565

If it is, maybe we should evaluate whether that auto-focus logic is worthwhile?

**Comment by youknowriad** [comment]
@aduth I don't it's related to #4872, it's more related to this comment https://github.com/WordPress/gutenberg/pull/4953#issuecomment-365245053 and this commit dbad7964284fd5f5a7a809f96ac27a2ebc8f9bec

It's a bit tricky though, because we want the empty paragraph block to always appear unselected while allowing these menus on hover.

File: editor/components/inserter-with-shortcuts/index.js
Comment (aduth) [inline editor/components/inserter-with-shortcuts/index.js]: Why `! isEmpty( item.initialAttributes )` ?

Comment (noisysocks) [reply editor/components/inserter-with-shortcuts/index.js]: It's so that shared paragraph blocks don't get filtered out.

Comment (aduth) [reply editor/components/inserter-with-shortcuts/index.js]: As a casual observer, I'd have no hints from the context available to reach this conclusion. Even now, I'm not clear what it is about shared blocks that causes us to filter them.

I'd suggest some combination of code comment and/or [explaining variable](https://refactoring.com/catalog/extractVariable.html).

**Pull request merged**
